Pm/device - initial fixes

ceolin opened the pull request:
- Do not wake up device unnecessarily
- Avoid unnecessary casts
- Keep subsystem only states private

erwango committed:
samples/boards: stm32: Add low power blinky sample

This sample demonstrates a low power blinky application.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/gpio: stm32: Enable PM_DEVICE services

Implement power mgmt hooks to support PM_DEVICE and
PM_DEVICE_IDLE.
In case of PM_DEVICE_IDLE, clock is requested for bank writes
so it is requested before configuring and released only if pin
is not configured as output.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/pinmux: stm32: Enable PM_DEVICE_IDLE services

Implement PM_DEVICE_IDLE using a workaround.
Calling device_pm primitives causes issues when running
tests/drivers/gpio.
Limitation: Prevents use of pinmux in PM_DEVICE_IDLE mode
in devices drivers.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers: gpio: Add support for stm32l562e_dk

Required for board testing


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

Commit:
power: device: Device pm state don't need to be exposed

It is confuse the power states used on device power management. There
are DEVICE_PM_STATE_* and DEVICE_PM_*_STATE states. The former is used
by the device pm subsystem and don't need to be exposed to the drivers.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device: Don't wake up devices unnecessarily

Currently when the system goes to sleep it asks *all* devices that
support PM to suspend or go to a low power state, and the system wakes
up it put *all* suspended devices in active state, even if a device
was already suspended and not being used.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device: void *context -> uint32_t *state

The context parameter used across device power management is
actually the power state. Just use it and avoid a lot of
unnecessary casts.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented:
Thanks @ceolin for working on this!
Some improvements, but i'm still seeing strange behavior when running `samples/boards/stm32/power_mgmt/blinky`
The LED (on GPIO G) is blinking, but some clocks are still unexpectedly resumed:

Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=true) 
> Switch on GPIO G on gpio_pin_configure_dt entry   -> OK
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=false)
> Switch on GPIO G off on pm_suspend               -> I'm not certain it should happen in DEVICE_IDLE (debatable)
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should not happen
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should really not happen
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=true)
> Switch on GPIO G on pm_resume                    -> should really not happen
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=false)
> Switch on GPIO G on pm_resume 
Breakpoint 3, gpio_stm32_clock_request (dev=0x200000ac <__device_dts_ord_19>, on=on@entry=true)
> Switch on GPIO G on pm_resume 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000014c <__device_dts_ord_138>, on=on@entry=true)
> Switch on GPIO B on pm_resume                    -> should not happen 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should not happen 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000014c <__device_dts_ord_138>, on=on@entry=true)
> Switch on GPIO B on pm_resume                    -> should not happen 
Breakpoint 3, gpio_stm32_clock_request (dev=0x2000016c <__device_dts_ord_137>, on=on@entry=true)
> Switch on GPIO A on pm_resume                    -> should not happen

Two issues:
- Still some wake up required on non used devices (GPIO A, GPIO B).
  - Not expected
  - Additionally, we could see 2 "clock on" w/o "clock off"  in between -> potential issue in the state machine
- pm_suspend and pm_resumed are invoqued on GPIO G
  - I'm not sure it should happen as device is supposed to handle its own state based on user request

gmarull commented:
Note: added myself to the review list, GH seems to be doing something weird with the reviewers list.

Commit:
power: device_pm: Check if is running in pre-kernel

Don't use kernel services before in pre-kernel state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pinmux: stm32: Use device_pm API

Use device runtime idle API now that this works in pre-kernel state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +575,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+		ret = gpio_stm32_clock_request(dev, false);
+	}
+
+	data->power_state = new_state;
```
state will be changed regardless of `ret` value

gmarull commented on subsys/power/device.c:
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
+	 * If the device is currently powered off or the request was
+	 * to go to the same state, just ignore it.
+	 */
+	if ((current_state == DEVICE_PM_OFF_STATE) ||
+			(current_state == state)) {
+		return false;
+	}
+
+	return true;
```
nit:

could be replaced by `return ((current_state != DEVICE_PM_OFF_STATE) && (current_state != state))`

gmarull commented on subsys/power/device.c:
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
+				LOG_DBG("%s did not enter %s state: %d",
+					dev->name, device_pm_state_str(state),
+					rc);
+				return rc;
```
in case of error, some devices will be left on `state` while some others will be kept on the current state. I wonder if this function should *undo* the state change in case of error.

ceolin replied:
Nope, the first device that fails will cause this function to return with an error code. The system pm then will undo the changes calling `pm_resume_devices()` that will only iterate over the devices that have changed state.

gmarull commented on drivers/display/display_st7789v.c:
```diff
@@ -409,14 +409,14 @@ static void st7789v_enter_sleep(struct st7789v_data *data)
 
 static int st7789v_pm_control(const struct device *dev, uint32_t ctrl_command,
-				 void *context, device_pm_cb cb, void *arg)
+				 uint32_t *state, device_pm_cb cb, void *arg)
```
that's a nice improvement. I wonder if states could be represented by an `enum`.

ceolin replied:
yep, I will work on states like we did to system power management. Just changed this because this signature was driving me crazy, every time I looked it I thought that context was some private data passing across callbacks but was just the state :/

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -177,6 +190,9 @@ void device_pm_enable(const struct device *dev)
 void device_pm_disable(const struct device *dev)
 {
+	__ASSERT(k_is_pre_kernel() == false, "Device should not be disabled"
+		 "before kernel be initialized");
```
nit:
`before kernel is initialized`.

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -112,6 +112,10 @@ static int device_pm_request(const struct device *dev,
 		}
 	}
 
+	if (k_is_pre_kernel()) {
```
I assume these changes imply that `pm_work_handler` will be never run in pre_kernel stage. Let's discuss that later more in detail.

ceolin replied:
yep, I tested with the blinky sample (last commit) and it seems to be working

gmarull commented on drivers/pinmux/stm32/pinmux_stm32.c:
```diff
@@ -151,8 +151,8 @@ int stm32_dt_pinctrl_configure(const struct soc_gpio_pinctrl *pinctrl,
 #ifdef CONFIG_PM_DEVICE_IDLE
 		/* FIXME: Use of device_pm clock mgmt function */
 		/* breaks correct gpio driver behavior */
```
`FIXME` comments likely need to go away now

ceolin replied:
yep, just changed to you be able to check / confirm

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -474,6 +474,12 @@ static int gpio_stm32_config(const struct device *dev,
 	}
 
+
+#ifdef CONFIG_PM_DEVICE_IDLE
```
I wonder if it would be better to just add *dummy* PM functions when device PM is not enabled, so that we do not pollute drivers with tons of `#ifdef CONFIG_PM_XXX`. See, e.g. https://elixir.bootlin.com/linux/latest/source/include/linux/pm_runtime.h#L231 Opinions @ceolin @nashif ?

ceolin replied:
I'm getting rid of CONFIG_PM_DEVICE_IDLE option, that should simplify all this `#ifdef` code. But I tend to agree with you about all these conditional code, we should avoid as much as possible. Dummy functions have to be carefully thought though, they will return an error code and these error codes have to be properly handled by the caller.

Commit:
power: device_pm: Use BIT macro for code compliance

BIT macro properly handle bit shift operations.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device: Remove pm_device_idle option

Simplify build options removing PM_DEVICE_IDLE. This is going to be
unified to PM_DEVICE. From devices perspective it does not change to
much, they are not enforced to implement or use anything.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Fix concurrence issues

The sync API was using k_poll_signal and in certain conditions is
possible multiple threads waiting on a signal leading to an undefined
behavior.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Remove device idle pm states

Avoid confusion with device runtime idle pm states and just use device
pm states.

This simplify the code a little bit and prepare the ground for having
a better definition of device pm states. Right now this code needed to
hijack two transitional states to not break the current code logic but
the goal is avoid it and have everything in one single place.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: device_pm: Add missing fallthrough

Add an explicit fallthrough in a case label.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +575,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+		ret = gpio_stm32_clock_request(dev, false);
+	}
+
+	data->power_state = new_state;
+
+	return ret;
```
should be changed to:
```c
if (ret < 0) {
    return ret;
}

data->power_state = new_state;

return 0;
```

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +575,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+		break;
+	default:
+		ret = -EINVAL;
+
```
nit: extra blank line not needed

gmarull commented on subsys/power/device.c:
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
+
+	rc = device_get_power_state(dev, &current_state);
+	if ((rc != -ENOTSUP) && (rc != 0)) {
+		LOG_DBG("Was not possible to get device %s state: %d",
```
nit: `LOG_ERR`

ceolin replied:
I can raise it to WRN but I don't think it should be ERR. ERR should be used for things that is not possible to recover

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -157,6 +161,26 @@ int device_pm_put_sync(const struct device *dev)
+		ret = device_set_power_state(dev,
+					     DEVICE_PM_ACTIVE_STATE,
+					     device_pm_callback, NULL);
+		if (ret == 0)  {
```
we should likely assert here, otherwise error condition will be silenced.

gmarull commented on doc/reference/power_management/index.rst:
```diff
@@ -492,10 +492,6 @@ the following configuration flags.
    This flag is enabled if the SOC interface and the devices support device power
    management.
 
-:code:`CONFIG_PM_DEVICE_IDLE`
```
Comparing to Linux, this flag enables what they call "runtime PM". It may still be useful to have a distinction between device PM (used by system PM) and runtime device PM. In some cases we may want to have one but not the other, e.g. I want that all devices are suspended when I suspend the sytem, but I don't care about PM when on running (all `device_pm_get/put` should be a nop). We can discuss this again tomorrow.

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -192,9 +191,9 @@ void device_pm_enable(const struct device *dev)
+		k_work_init_delayable(&dev->pm->work, pm_work_handler);
 	} else {
-		k_work_submit(&dev->pm->work);
+		k_work_schedule(&dev->pm->work, K_USEC(0));
```
`K_NO_WAIT`

ceolin replied:
will change, that is the part that I'm not particularly happy, I have to use `k_work_delayable`  otherwise the thread is scheduled out and the notification is missed causing a dead lock. For now this seems to work, but I want to remove the need of the work queue.

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -116,25 +121,19 @@ static int device_pm_request(const struct device *dev,
 	}
 
-	k_work_submit(&dev->pm->work);
+	(void)k_work_schedule(&dev->pm->work, K_USEC(0));
```
`K_NO_WAIT`

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -207,6 +206,6 @@ void device_pm_disable(const struct device *dev)
 	dev->pm->enable = false;
 	/* Bring up the device before disabling the Idle PM */
-	k_work_submit(&dev->pm->work);
+	k_work_schedule(&dev->pm->work, K_USEC(0));
```
`K_NO_WAIT`

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -8,6 +8,7 @@
 #include <kernel.h>
 #include <device.h>
 #include <sys/__assert.h>
+#include <sys/check.h>
```
why is this needed?

ceolin replied:
leftover from a check I had added but end up removing.

gmarull commented on subsys/power/device_pm.c:
```diff
@@ -18,27 +18,19 @@ LOG_MODULE_DECLARE(power);
+ * transitioning states for internal usage.
+ */
+#define DEVICE_PM_RESUMING_STATE   (DEVICE_PM_OFF_STATE + 1U)
+#define DEVICE_PM_SUSPENDING_STATE (DEVICE_PM_RESUMING_STATE + 1U)
```
I'd not define internal states based on the value of external ones, a change on the external side may create unwanted value clashes.

ceolin replied:
I will unify this with that states in near future. But there was already a dependency between those states, but was very confuse having `DEVICE_PM_ACTIVE_STATE` and `DEVICE_PM_STATE_ACTIVE` :/

gmarull reviewed (commented):
Added some more comments/questions

nashif commented on subsys/power/device.c:
```diff
@@ -60,26 +60,65 @@ static device_idx_t num_pm;
+		 * between other states like DEVICE_PM_LOW_POWER_STATE.
+		 */
+		if (suspend) {
+			++num_susp;
```
why not fold into previous if?

ceolin replied:
There is no reason looking it. Will put it inside the previous branch.

Commit:
power: device_pm: Use spin lock instead of semaphore

Device pm runtime was using semaphore to protect critical section but
enable / disable functions were waiting on the semaphore. So, just
replace it with a spin lock.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1583,7 +1583,7 @@ DEVICE_DT_INST_DEFINE(index,						\
 		    &uart_stm32_pm_control,				\
 		    &uart_stm32_data_##index, &uart_stm32_cfg_##index,	\
-		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
```
@gmarull Yes. But this change is not meant to be merged. The only purpose was to be able to have a working configuration the time the underlying limitation is fixed. I think we'll soon be able to get rid of it.

Commit:
power: device_pm: Power device on/off in pre-kernel

Devices may be initialized but started powered down for this reason
is necessary to power a device on if requested even if in pre-kernel
state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
